tiledb/sm/consolidator/fragment_consolidator.cc

```diff
@@ -542,9 +547,14 @@ Status FragmentConsolidator::create_queries(
   RETURN_NOT_OK((*query_r)->set_layout(Layout::GLOBAL_ORDER));
 
   // Refactored reader optimizes for no subarray.
-  if (!config_.use_refactored_reader_ ||
-      array_for_reads->array_schema_latest().dense())
+  if (!config_.use_refactored_reader_ || dense) {
     RETURN_NOT_OK((*query_r)->set_subarray_unsafe(subarray));
+  }
+
+  // Enable consolidation with timestamps on the reader, if applicable.
+  if (config_.with_timestamps_ && !dense) {
+    (*query_r)->set_consolidation_with_timestamps();
```

Comment:
Maybe I didn't get that right, but shouldn't we enable this on the writer as well so that `has_timestamps` in `WriterBase` below is `true` in consolidation case? I couldn't find how it would be true otherwise.

Reply:
The writer can use the presence of the __timestamps buffer. The reader cannot do that since requesting the __timestamp values is a valid scenario.